### DISCUSSION TITLE
vAmiga web ports

### DISCUSSION THREAD

**Discussion by mithrendal** [vAmigaWeb/vAmigaWeb#1] [Announcements]
Hi @dirkwhoffmann  @chris70c 

can you both hear me ? 🙃

@dirkwhoffmann  we have to provide some newer pictures of vAmiga mac UI here for chris, because chris does not have a mac and he likes to reimplement  inspector and all that fancy stuff from the vAmigaMac in his web version ...

TASK1:  add some newer pictures of vAmiga for chris here ... 😌

**Comment by mithrendal** [comment]
@chris70c maybe this story is interesting for you ? 

https://github.com/dirkwhoffmann/vAmiga/issues/291

it is about bringing vc64web to live and how we struggled in the beginning with pthreads ... and how we solved it ...

Can you tell us why you did choose in your current implementions, the pthread approach over the wasm singlethread appoach ? As I understand, is that it is not (yet) supported by all browsers ? Does it have other advantages ?

**Comment by chris70c** [comment]
Hi all, let's see...

First consideration was "leave the original code as it is as much as possible", I've changed only 4 files, Amiga.cpp,
Amiga.h, Thread.cpp and Thread.h (there other small changes around but those are only to remove any reference to
RegressionTester, RetroShell, Recorder and the Script file format).

The advantages are quite honestly big, the emulator code runs on a separate thread from the main dom/js thread, meaning
that for example the debugger functions can be more easily implemented and will not have effect on the emulator even
when transferring data to/from js/c++.

SharedArrayBuffer have been around since 2018, they were removed for a short period of time from all browsers to mitigate
the effect of Spectre/Meltdown vulnerabilities but they've been back in all browsers (but Safari) since at least half
a year now, it only requires your server to send the appropriate CORS headers which is quite a simple thing to do and
it does not affect the end user. Hopefully Apple will decide to enable it back behind CORS, like everyone else did,
soon enough. (Edge, Chrome, Firefox, Chrome for Android, Firefox for Android, all supports them).

The last reason is that at the moment I'm using a modified SDL2 version (which completely removes all the dom listeners
automatically added which I don't need at all and just make my life miserable) but SDL doesn't support the new web audio
api (indeed you get a deprecated message in the console each time), so I plan to write my own audio backend (based on the
work already done for FLOD that now runs with the latest web audio api), this will also improve the emulator performance
because the audio runs on a different thread leaving more time to run the regular dom/js code but to do that I need access
to the raw memory buffer used by webassemly and SharedArrayBuffer is the only solution for that.

Both ports of vAmiga and VirtualC64 have been modified in the same way and require SharedArrayBuffer to work.
BTW, quite interesting reading the link above, thanks.

PS: I don't have anything against Apple but Safari is underwhelming at best, they choose what they like of the standards and
leave the rest out, vAmiga wouldn't work on iOS anyway 'cause the PointerLock api is missing and without it there is no other
way to move the Amiga mouse cursor, you could hide the OS cursor but you would loose the mouse as soon as the cursor
goes outside the canvas rectangle.

**Comment by mithrendal** [comment]
I just tried my old dmcs.adf .... on my very old macbook pro from 2007

![image](https://user-images.githubusercontent.com/17108995/133060625-9c5d906b-4fce-4eb3-b81b-e81277ed3d7f.png)

it plays the song "La bomba" fluently without any hicups ... so nice ... @chris70c  my old rusty macbook is internally known as the grandpa mac and dirk was already successfully to optimize the performance of the emulator core to let it execute vc64web without getting a heart attack ... (it gets very hot)

**Comment by mithrendal** [comment]
Very exciting and cool approach to make the use of pthreads with web workers and SharedArrayBuffer 🤓

I took a look on the grand pa mac with latest firefox 78  which I got some weeks ago from the ESR update channel and it gives me ...
Uncaught ReferenceError: SharedArrayBuffer is not defined
vamiga.js:1:6604

I cannot upgrade to any newer firefox on the grand pa 😬 ... but at least 😄 it works well in the latest chrome see picture above ... 

so it seems ... the system requirements for the pthread version of vAmiga in the web will be a bit more particular, 
  needs latest browsers
  does not work on any iOS device due to safari 

I plan to bring a vAmiga web version which is a takes the current http://vc64web.github.io as a template ... which in turn runs on any device with any browser not older than 4 years ...  as it goes the single threaded approach ... which like you said requires more invasive modifications of dirks core ... which in turn is quite a bit annoying ...

we will have two versions then, one for the power user (yours) who wants to see all the glitter and fancyness of the Amiga machine and another for the rest ... which maybe only want to game ... and do some AI scripting ... AI scripting for example is a thing where  http://vc64web.github.io is very strong ... it allows you to write AIM Bots for games with which you can beat the games internal AI ... (see https://github.com/vc64web/virtualc64web/wiki/javascript) ... you reach this by tapping the + sign in the navbar ... and it allows you to write little action scripts and link them to buttons which in turn can take over control of joystick ports or other ports as well ...

technically the two version approach will be very good for both projects as we can learn from each other ...

**Reply by chris70c** [reply]
You can check this page to find out exactly which browser supports what:
https://caniuse.com/?search=sharedarraybuffer

The more the merrier :)
3 versions each, no p-thread, p-thread, developer edition, lol

**Comment by mithrendal** [comment]
>you can check this page to find out exactly which browser supports what:

oh from Firefox 79 upward   ... so close ... with grand pa I am stuck on 78 😫 ... luckily he still knows how to execute chrome 93

>The more the merrier :)
>3 versions each, no p-thread, p-thread, developer edition, lol

🤔 what do you mean with developer edition ? did I miss something ? or do you mean the native Mac version?

**Comment by chris70c** [comment]
Developer edition is what I call the web version with the full Inspector interface (and some other things like module ripper, gfx ripper, sprites viewer, etc.)

**Comment by mithrendal** [comment]
@chris70c you managed to run big sur and xcode ... then why don't you just run it to see the inspector?🤔 you even do not need xcode ... just download the latest version and run it ... no?



🤓wait I will capture some screenshots for you and put them here in a minute or so... and answer your questions from your mail

**Reply by chris70c** [reply]
Forgot to tell you, no Windows VM supports Metal so I can only look at the code but I can't run anything, lol

**Reply by mithrendal** [reply]
Oh I see ... 🤔the best would be maybe a movie ... @dirkwhoffmann what about another youtube video? which demonstrates all the cool things in vAmiga for chris ?😎

**Reply by chris70c** [reply]
I've updated both emus online with webgl shaders now, I'm gonna add at least the rom config to it then I'll upload the code to github, thanks for the answer, I will start making all the ui for the inspector then I will ask any question, I'm using embind for this so I got all the classes available in js, the cpu registers and part of paula and denise are already working.
The only problem with embind is that I can't export raw pointers to primitive types so for example the ScreenBuffer class which as a data property pointing to a u32 array can't be exported as it is; I will think about something for those.

**Reply by mithrendal** [reply]
take your time ... this sounds all very promising 😍 ... when you have your repo online, we can also post screenshots and answer questions there ...
I guess I don't use embind, instead I use the simpler emscriptens cwrap to bind c functions, which only know strings and numbers as parameter as well ... when I like to export something else to JS (other than numbers or string) then I came up with the idea to use json strings, which I set up in the mainsdl.cpp class. With them you can transport anything ... I am not sure if that solution suits to your u32 array too? Some kind of a JSON-Proxy in between ... maybe too slow?

**Comment by mithrendal** [comment]
![image](https://user-images.githubusercontent.com/17108995/133147020-dc9c00c2-1d83-4f9b-8858-09e30091fc74.png)

leftmost imagebutton at the bottom of inspector panel is "pause/run" toggle, then comes "step into", and the third button is "step over".

you can interact with some of the text inputs but not with all of them, the checkboxes seem to be for visual representation of the state only ... @dirkwhoffmann ? correct ? this is what I see in the v1.1 at least when I click around at insert workbench hand picture....

when you press run then the visual state of all the inspectors will change ... for example at CPU sub panel the program assembler code begins to scroll...

**Comment by chris70c** [comment]
Agnus and Paula Inspector panel

![vamigadev](https://user-images.githubusercontent.com/3858768/133303310-d7c7932a-ba4c-4529-97e2-8a7031058c2d.png)

same with different Flesk theme (Amiga theme):

![vamigadev(amiga)](https://user-images.githubusercontent.com/3858768/133306735-9d2dd82d-2d7f-4937-8322-9c143e05bdf2.png)

**Reply by mithrendal** [reply]
that looks nice ... are these panels movable ?

**Reply by chris70c** [reply]
Yes Flesk is my desktop alike framework, those are real windows, resizable, minimizable, maximizable and movable :)
The entire desktop is also completely themable and scalable (see that 1.00x at the right top of the window, that's the scale button). You won't like the fact that is Chrome/Edge only thou, lol, unfortunately I'm using all the latest (including adoptedStyleSheets with import and asserts) and the other browser are so way behind (particularly with Houdini apis) that if you really want to explore something new you have to leave them behind.

**Comment by mithrendal** [comment]
@chris70c just a word to your issue findings 😎 ... keep on pushing new issues to the vAmiga repo ... I know dirk will be very happy about those findings ... he is just not responding because he is currently offline ... 🤓

In the meantime, you can try different blitter accuracy settings ... because vAmiga has different blitter implementations (I think 3) ... you can configure them



>I guess I don't use embind, instead I use the simpler emscriptens cwrap to bind c functions, which only know strings and numbers as parameter as well ... when I like to export something else to JS (other than numbers or string) then I came up with the idea to use json strings, which I set up in the mainsdl.cpp class. With them you can transport anything ... I am not sure if that solution suits to your u32 array too? Some kind of a JSON-Proxy in between ... maybe too slow?

was the idea of a json proxy bridge of use for you ?

**Reply by chris70c** [reply]
No problem with that no worries, I have a friend who's doing most of the testing, he's having fun :)
I'm not clear about ECS support thou, I saw there are 2 configurations at the moment, OCS 512K+512K and ECS but still 512K+512K, most ECS configurations are with 1MB chip and I saw some comments about ECS not being ready yet so for now I'm only testing against WinUAE quickstart Amiga500 OCS 512K+512K with KS 1.3.

As for the json bridge I haven't try anything yet, just working on designing most of the windows for the moment but I'll look into it for sure, the only problem may be speed, not sure is gonna be fast enough to share all the data when debugging live, but we'll see.

**Comment by chris70c** [comment]
When you have time if you can post a screenshot of the Denise panel would be great, I can see that there is some text in the white box under sprites but I have no idea if it is in columns or just a line of text for each row or something else, lol

![denise](https://user-images.githubusercontent.com/3858768/133348719-67067a43-926b-4eee-8a38-c04e1712827b.png)

**Comment by chris70c** [comment]
Another panel and some tweaks to the others, lol

![test3](https://user-images.githubusercontent.com/3858768/133351786-d68947cd-9d7f-4c2f-8548-eebf99db6791.png)

**Comment by mithrendal** [comment]
![image](https://user-images.githubusercontent.com/17108995/133405028-5f70a7f0-94d4-4ab1-a761-7843f1431adf.png)

**Comment by chris70c** [comment]
Updated devEd online, also added theme switcher (dropdown in the desktop menu).
I have a question about the CPU disassembler control, does it scroll vertically or only horizontally?

**Comment by dirkwhoffmann** [comment]
The CPU disassembler window stretches vertically (it also scrolls horizontally if the disassembled line is too large):

<img width="980" alt="Bildschirmfoto 2021-09-18 um 10 17 00" src="https://user-images.githubusercontent.com/12561945/133883859-2725fa87-93e4-4b16-90ee-e8639bf0f595.png">

Here is a screenshot of the trace window:

<img width="980" alt="Bildschirmfoto 2021-09-18 um 10 17 05" src="https://user-images.githubusercontent.com/12561945/133883873-a052e621-d804-4f24-b926-b25fe60b7020.png">

**Reply by chris70c** [reply]
Ok, so from what I can understand from the swift code it looks like you're disassembling about 256 instructions at a time, but there is no vertical scrollbar, so how do you look at the code past what fits in the window and how do you decide when to disassemble more code?

**Reply by chris70c** [reply]
Maybe a small video would help if its not too much trouble.

**Reply by chris70c** [reply]
Working on a small component for the disassembled code:
https://www.neoartcr.com/Flesk%201.00/

**Reply by mithrendal** [reply]
I like the way how the window component feels nearly native ... 🤤

**Comment by dirkwhoffmann** [comment]
> Updated devEd online, also added theme switcher

The theme switcher stopped working (I think). It is greyed out (Chrome browser).

**Reply by chris70c** [reply]
Temporarily disabled, need to do some restyling and I didn't have time to add the new windows done yesterday night, will be back soon :)

**Reply by chris70c** [reply]
Theme switcher enabled back with new Amiga theme styling.

**Comment by dirkwhoffmann** [comment]
https://user-images.githubusercontent.com/12561945/133918120-fd787b55-cd81-45de-bd70-8343ef71d05b.mov

**Comment by dirkwhoffmann** [comment]
The CPU inspector works as follows:
- The currently executed instruction is highlighted by selecting the corresponding row in the NSTableView.
- If the next executed instruction is already visible, only the selected row is changed.
- If the next executed instruction is not yet visible, the disassembler window is recreated with the new PC in the first row.

**Comment by dirkwhoffmann** [comment]
Here is the corresponding Code (InstrTableView.swift):

```
    func jumpTo(addr: Int) {

        if let row = rowForAddr[addr] {

            // If the requested address is already displayed, we simply select
            // the corresponding row.
            reloadData()
            jumpTo(row: row)

        } else {

            // If the requested address is not displayed, we update the data
            // cache and display the address in row 0.
            cache(addrInFirstRow: addr)
            reloadData()
            jumpTo(row: 0)
        }
    }
```

**Reply by chris70c** [reply]
Got it, so there is a vertical scrollbar there in the video, I guess that's to scroll the 256 instructions loaded at any time?

**Reply by dirkwhoffmann** [reply]
Yes, exactly.

**Reply by chris70c** [reply]
Ok, last couple of questions about this:

What this button do? (refresh?)
![133883859-2725fa87-93e4-4b16-90ee-e8639bf0f595](https://user-images.githubusercontent.com/3858768/134253940-d2dde955-686d-4fa0-af62-db194ba0dc2f.png)

The search I assume is for instructions (like search for rts)?

**Reply by dirkwhoffmann** [reply]
Yes and yes. The button triggers a full refresh. It is not really needed (if everything works as expected). Maybe I should remove it.The search field takes an address and starts disassembling from this address. If the address is already in the cached, it simply jumps to the corresponding line.

**Comment by chris70c** [comment]
https://www.neoartcr.com/vamiga/

New look and (not working yet) configuration screens :)

**Reply by mithrendal** [reply]
🤤 🤤 🤤

**Comment by chris70c** [comment]
Can you show me a screenshot of the Configuration Compatibility page when the emu is running, is there any option there that should be disabled (like in Hardware) ?

**Comment by mithrendal** [comment]
![image](https://user-images.githubusercontent.com/17108995/134629455-9ea601bd-9227-44e8-84b0-3bdb5028863e.png)

you can change everything on the fly while running

**Reply by dirkwhoffmann** [reply]
Mithrendal, thanks for posting.

All compatibility settings have been design to be "hot pluggable". You can change them any time (even the Blitter setting. vAmiga ensures that an ongoing Blitter operation is finished at the same accuracy level as it was started).

**Comment by chris70c** [comment]
Some love for VC64 too:
![c64_roms](https://user-images.githubusercontent.com/3858768/134748859-7c26b1d2-ded7-4d29-9c60-31a2e49514f2.png)
![c64_hardware](https://user-images.githubusercontent.com/3858768/134748862-3b4019e2-dac9-4337-958b-a6e4a763f83a.png)
![c64_peripherals](https://user-images.githubusercontent.com/3858768/134748865-09318b0a-0727-4f2d-9011-96412314bd9e.png)
![c64_compatibility](https://user-images.githubusercontent.com/3858768/134748869-31d7d50a-d904-482f-983d-09b363c10e6d.png)

**Reply by dirkwhoffmann** [reply]
Gorgeous. One needs to look twice to see that it's not the Desktop version.

**Reply by chris70c** [reply]
Final configuration panel for Virtual C64:
https://www.neoartcr.com/virtualc64/UI/

Final configuration panel for vAmiga:
https://www.neoartcr.com/vamiga/UI/

**Reply by dirkwhoffmann** [reply]
The panels are great!

Two minor things: 
- When viewed in Safari, the popup buttons carry two arrow symbols which overlap each other:
<img width="212" alt="Bildschirmfoto 2021-09-26 um 08 00 35" src="https://user-images.githubusercontent.com/12561945/134795551-4da977d6-d2c6-4fe3-a250-2251763cfe2d.png">
- "Revert..." does not contain any items (yet).

**Reply by chris70c** [reply]
Ah, I see, looks like Safari doesn't let you remove the default appearance of the dropdown box, I'll see if I can override the styling and leave the original arrows for Safari only; Yes, revert doesn't contain any item yet, I'm looking in the xib file to see what options you're setting up for the different panels. I'm not 100% sure that all the options will be possible in the web version but we'll see.

**Reply by chris70c** [reply]
I have a question about the knobs in the audio and video panels.

The audio ones are used for panning, from what I can see in the code (VC64) the value goes from 0 to 200 (included), are you using only the upper 180deg so that 0 is at -90deg (aligned to the letter L) and 200 is at +90deg (aligned to letter R) or are they using the full 360deg span?

What about the ones in the video panel?

As always a small video would be the best :)

**Reply by chris70c** [reply]
Fixed some bugs in Safari (hopefully).

**Reply by dirkwhoffmann** [reply]
Dropdown boxes in Safari are fine now 👍:

<img width="283" alt="Bildschirmfoto 2021-09-28 um 07 52 20" src="https://user-images.githubusercontent.com/12561945/135031343-7bc96b7c-c7d6-468b-83f9-ce4ae313beef.png">

**Reply by dirkwhoffmann** [reply]
Here is a video for the video panel (recorded with VirtualC64):

https://user-images.githubusercontent.com/12561945/135031560-d6fda2cf-b0bf-4dfb-b781-30e63a577085.mov

**Comment by dirkwhoffmann** [comment]
Hi Chris,

here is the relevant code for the pan setting:

```
       case OPT_AUDPAN:
                        
            assert(id >= 0 && id <= 3);
            
            config.pan[id] = std::clamp(value, 0LL, 200LL);
            
            if (value <= 50) pan[id] = (50 + value) / 100.0f;
            else if (value <= 150) pan[id] = (150 - value) / 100.0f;
            else if (value <= 200) pan[id] = (value - 150) / 100.0f;
            return;
```

The valid value range is 0..200. In the code shown above, the values are mapped to interval [0 .. 1] where 0 means "all left" and 1 means "all right" (or vice versa). The code has been done this way, because the GUI uses circular sliders which means that for most pan settings, there are two possible slider positions. E.g., 45° and 135° are the same. 

I don't like the current implementation very much, because the design of the emulator logic is driven by the GUI. Usually, it should be the other way round.

Today, I'll be pretty busy. I'll do some videos (about the video panel) when I'll have some spare time...

**Reply by dirkwhoffmann** [reply]
I've reworked the audio pan settings. The new (official) value range is -100 ... 100 where -100 means "left speaker only", 0 means "centered audio", and 100 means "right speaker only". Internally, the value is mapped by this formula:

    pan[id] = 0.5f * (sin(config.pan[id] * 3.14159f / 200.0f) + 1);

This formula maps [-100; 100] to [0;1]. Using the sine function has the nice side effect that it's easy to support the circular sliders in the GUI panel. We can simply set the slider's value range to [0;400] to get the correct effect.

**Reply by chris70c** [reply]
I wasn't happy with the knob so I wrote a new control:
https://www.neoartcr.com/knob/

**Reply by dirkwhoffmann** [reply]
The new knob is great. It's much better than the one we have in macOS 👍.

**Comment by dirkwhoffmann** [comment]
All Rom database extensions made by @chris70c have been merged in. I've also adjusted the GUI to display the new model information:

<img width="886" alt="Bildschirmfoto 2021-10-01 um 15 53 06" src="https://user-images.githubusercontent.com/12561945/135632369-6f76bb4b-07a7-463c-8a93-1b277f599765.png">

<img width="886" alt="Bildschirmfoto 2021-10-01 um 15 53 17" src="https://user-images.githubusercontent.com/12561945/135632374-94516f3a-0e67-49e0-93ac-87757eea778d.png">

**Comment by chris70c** [comment]
New version online:
https://www.neoartcr.com/vamiga/

**Reply by mithrendal** [reply]
I tried with chrome... it says NET::ERR_CERT_DATE_INVALID 


![image](https://user-images.githubusercontent.com/17108995/135749526-71026fcc-313e-4975-ab0a-5c07102fe902.png)

**Reply by chris70c** [reply]
No idea what can be wrong, its complaining about the ssl certificate (it has nothing to do with the emu) but if it was working before I don't see why it stop. My ssl certificate is fine as you can see from here:
![ssl](https://user-images.githubusercontent.com/3858768/135756521-b2c72dcd-f911-4da9-a561-c8631043e046.png)

**Reply by dirkwhoffmann** [reply]
On my machine, it works well with Chrome:

<img width="724" alt="Bildschirmfoto 2021-10-03 um 16 02 47" src="https://user-images.githubusercontent.com/12561945/135757211-cba545a4-b71d-4fab-882e-debeb99a46b3.png">

**Comment by dirkwhoffmann** [comment]
Unsure if I encountered a bug in the web version. When loading Agony in the Desktop version, the screen becomes black and a short moment later, this comes up: 

<img width="886" alt="Bildschirmfoto 2021-10-03 um 16 05 35" src="https://user-images.githubusercontent.com/12561945/135757311-7603f279-d8aa-42ea-9d12-83701e11444f.png">

When I run the game in the web version, the screen remained black.

**Reply by chris70c** [reply]
Works fine with this config (I haven't set them all yet by default)
![agony2](https://user-images.githubusercontent.com/3858768/135758603-e705edc3-ef34-4d99-b043-a44ff342e8cb.png)
![agony1](https://user-images.githubusercontent.com/3858768/135758607-60ef0a03-9caf-4aa5-a11d-5b8cbeeac6e4.png)

**Comment by mithrendal** [comment]
![image](https://user-images.githubusercontent.com/17108995/135761451-49274227-abe1-474b-a9e0-f36ff344cd99.png)
it complains on my chrome that the root cert was only valid to 30.september 2021 that is 4 days ago ....


strangely it works on firefox ... I guess my chrome has still got the old root cert of your signing authority ...

**Reply by chris70c** [reply]
Maybe you catch it exactly while it was renewing the certificate, Let's Encrypt is a free SSL service but it requires to renew the certificate every 30 days; the server does it automatically but I guess it pass some time before the new certificate is deployed and recognized.
Hopefully it will fix itself soon...

**Reply by mithrendal** [reply]
I could bypass in the meanwhile with a "connect anyways..." option, 🤤  (give me access to the thing even if it is dangerous .... haha!)


![image](https://user-images.githubusercontent.com/17108995/135764265-1f298818-0065-44ba-8393-7055a0955cce.png)
 
and it loads one of my favourites ... joystick with keyboard worked for directions but what is the fire key ? 🤔

@chris70c It works so nice !! 😍 Speed is okay too !!!... on my old grandpa ... 

you have done an incredible good job here !!! I like the rom dialog !!!

**Reply by chris70c** [reply]
Thanks, hopefully I'll be able to add all the remaining features soon.

Joystick keyset in port 1 is:
Left = A
Right = D
Up = W
Down = S
Fire = Left Control  or  B

Joystick keyset in port 2 is:
Cursor keys for moving
Fire = Right Control  or  (slash key) /

**Comment by chris70c** [comment]
Added Cloanto roms support (decrypted in js)
![romkey](https://user-images.githubusercontent.com/3858768/135765938-23a3580a-dfd8-4728-b747-be863cc839d6.png)

It looks like they removed the rom.key file from the latest versions of Amiga Forever unfortunately, I've installed 9.2.6.0 and I can't find it anywhere :(

**Reply by chris70c** [reply]
Yep, they removed the rom.key, changed the encryption and added a ROM Toolbox tool to the player which can save the rom decrypted. The rom.key I have (version 10.0) works for the old roms but not for the new ones, not sure from which version they changed it, need to install each version until I found one which still has the rom.key.

**Comment by chris70c** [comment]
Update VC64 online (https://www.neoartcr.com/virtualc64/), still early beta but something works, lol

**Comment by dirkwhoffmann** [comment]
Three small videos showing how to add a T64:

1. Insert the T64 as a disk by using a drop zone

https://user-images.githubusercontent.com/12561945/136513550-96189d2b-bbd3-4658-a63a-d22c0a57624a.mov

2. Insert a T64 as a disk by using the disk dialog

https://user-images.githubusercontent.com/12561945/136513609-ddbe0a60-10e1-4e77-8bc0-0af252bf50df.mov

3. Flashing a T64 into memory

https://user-images.githubusercontent.com/12561945/136513779-10d852bb-58b2-4c83-ab55-6c7203ab57e3.mov

**Comment by chris70c** [comment]
Snapshots screen...
https://www.neoartcr.com/snapshots/

**Comment by chris70c** [comment]
Hi everyone, been some time :)
I've put online the demo of the new UI for vAmiga 2.0 here: https://www.neoartcr.com/ui_test/

Changes are:
  new portrait/landscape mode for the emulator window
  new light/dark theme mode
  update configuration panel for vAmiga 2.0 including HD support
  new fullscreen mode
  screen with a 1001 hd containing whdload installed games ready to use (rocket button)  :)
  updated light theme colors with better contrast and redone most custom controls to be more user-friendly

that's all, hopefully the full version will be ready soon...

**Reply by mithrendal** [reply]
😍 that looks really so good chris. I have also found Archon within the game browser. The UI is fantastic. At the moment it is only a mockup of the ui right? I could not power up the vamiga. Can’t wait to try the real thing. 🫶 mille grazie

**Reply by dirkwhoffmann** [reply]
Excellent work! I am also looking forward to the final release.

**Reply by chris70c** [reply]
Beta 2 online:

Changes are:
switched to my new framework, no more modal dialog windows, now each window behave like a real desktop window, you can drag them around, they have appropriate z-index layering when clicked, etc. you can move from one to another with alt+left/right cursor
minor cosmetic changes
update virtualc64 to be on par with vAmiga
fixed fullscreen layout in Firefox
added back improved Snapshots and About windows

new links:
https://www.neoartcr.com/ui_test/vamiga
https://www.neoartcr.com/ui_test/virtualc64

![overview](https://user-images.githubusercontent.com/3858768/162340179-433bf059-5219-450a-ad07-31bb59ffa342.png)

![overviewc64](https://user-images.githubusercontent.com/3858768/162340317-7b7606d8-1572-42e6-8482-315c4f7b8d24.png)

**Comment by chris70c** [comment]
Test version is online, still plenty of missing stuff but it's a start :)
https://www.neoartcr.com/ui_test/vamiga
There's only 1 whdload game online at the moment, 4x4 Off Road Racing...

**Reply by mithrendal** [reply]
Kalimera 🫡 still in holidays. On a greek island… already impatient but I have got no computer here only iPhone 😖 … will watch your masterpiece next week 😎

**Comment by mithrendal** [comment]
just tried it ... beta3 ... I was stunned by the user interface ... looks really good ... however I could not run 4x4  off road ... first I tried to just click on the checkmark and then it seems to do a reset and nothing booted ... e.g. disk hand screen ... then I thought maybe it needs a workbench floppy to so that I could see the hd0 ... but also no success ... anyway it was a great experience so far ... when it is fully completed it will be very amazing stuff ...

**Comment by chris70c** [comment]
You need to manually change the config at the moment, 2M Chip + 8M Fast + HD0 controller, I still don't have in place a way to automatically load a whdload working config.

**Comment by chris70c** [comment]
The new version if officially online here:
https://www.neoartcr.com/vamiga
games are available from A to I included for now.
First think to do is install a KS 2.04 rom otherwise nothing will work (config is automatically loaded now), I still need to find a better way to do this but without loading the KS from the server (which I can't do for copyright issues) there is no easy way to handle it.

**Comment by dirkwhoffmann** [comment]
> The new version if officially online here:

Hi Chris, great work! I played "Adams family" this morning for quite some time in Firefox without any issues 😎.

BTW, this image is great 😊: 

<img width="552" alt="Bildschirmfoto 2022-04-28 um 14 00 47" src="https://user-images.githubusercontent.com/12561945/165747702-b10e444f-b796-4f58-b185-7b7971cf3a49.png">

**Comment by mithrendal** [comment]
I played 1943. 😎It was easy. Wonder if it has been trained because normally these old games are quite difficult in general

**Comment by dirkwhoffmann** [comment]
There is no Defender of the Crown, isn't there? 🤧 No worries, I've played "Die Hard 2: Die Harder" instead 😎. It's really difficult (at least for me), but it's still fun.
 
The game gallery is really really cool. It's like an online retro museum with immediate access to all what we'd liked back in the day.

**Comment by chris70c** [comment]
no Defender of the Crown because it requires a KS image on the hdf, for the moment I've decided to put online only games that do not require any KS image on the hard disk, I have about 601 of them ready, later on will see if I don't get any email to take everything down I'll think about it, asked Cloanto but I got no answer :(

**Comment by chris70c** [comment]
Uploaded games J-Z (all games are online now!)

**Comment by mithrendal** [comment]
> ...asked Cloanto but I got no answer :(

chris they never answer ... I have asked them for permission to include the commodore amiga system library documention into visual studio code extension "amiga assembly" but they never answered ... dirk asked them something about permission ... no answer ... seems that this company doesn't like to communicate 🙄 maybe they all are busy with real jobs and cloanto is just a spare time hobby for them ...

**Reply by chris70c** [reply]
Yes, I've already tried 5 or 6 times in the last 2 years, lol, but they never answer, too bad 'cause if we could include at least one image of a KS2.04 it would be much easier, you can load games directly from the url (the format is ?id=0,1943; where 0 is the index of the page (0 = Games) and the name of the archive (1943)) but if you don't have a suitable KS image installed it won't work (I will add some message later on)...

**Comment by chris70c** [comment]
VirtualC64 0.9 Beta (all games are online)
https://www.neoartcr.com/virtualc64_beta/

needs all roms (1541 included for save disks)

**Comment by mithrendal** [comment]
chris I played in yours virtualC64 ACE 🤤 🤤 🤤   ... I still remember this one ... it was my favourite on 7. july 1985 ... when boris becker won wimbledon the first time ... everyone was looking tennis ... me playing ACE 😎

**Comment by chris70c** [comment]
Miracles do happens sometime :) I got an answer (useless but an answer from Cloanto):

Hello Christian,

Thanks for reaching out.

Right now server-side licenses are limited by existing agreements. However, this may change on a year by year basis, and once the new Amiga takes over I hope that new options become available. I took a note of your request for later contact.

Please do not hesitate to let me know if I can be of further assistance.

Best regards,
Takeo


What's the new Amiga ??????

**Comment by dirkwhoffmann** [comment]
> What's the new Amiga ??????

I was just about to ask you that 😂.

**Reply by chris70c** [reply]
Absolutely no idea, a new company? Cloanto changing name 'cause they bought all the remaining rights to the Amiga name?
Who knows, Amiga history after Commodore is quite complicated, not sure who actually owns what about it, lol.

Update VirtualC64 beta, auto swap joystick ports when a game is loaded, minor fixes here and there...

**Comment by mithrendal** [comment]
Chris what have you asked them ? .. permission to bundle kick20 ?At least a friendly answer I think... and they offer further assistance... hm with what?

**Reply by chris70c** [reply]
original message from me:

-----Original Message-----
From: Christian Corti ([christian@neoartcr.com](mailto:christian@neoartcr.com)) [mailto:donotreply-v1-
[web@cloanto.net](mailto:web@cloanto.net)]
Sent: Thursday, April 28, 2022 15:16
To: RetroPlatform Help Desk [<support-rp-v3@cloanto.net>](mailto:support-rp-v3@cloanto.net)
Subject: Kickstart roms for online use in emulator

************************************************************
Type:         Licensing and Sales
Subject:      Kickstart roms for online use in emulator
Name:         Christian Corti
Organization:
Email:        [christian@neoartcr.com](mailto:christian@neoartcr.com)
Contact:      https://www.neoartcr.com/
Form URL:     https://www.amigaforever.com/contact/
Time:         2022-04-28T15:15:55Z

Hello, I'm the author of the online port of vAmiga (an alternative to WinUAE
for Amiga emulation), at the moment I have to ask users to install their own KS
rom given the copyright nature of it, unfortunately this will make the process
not quite user friendly as I would like it to be, lots of users have no idea  what a
KS rom is or where to find it... Would it be possible to discuss some kind of
agreement for the rights to use at least KS 1.3 and 2.04 (online use only, no
download of course)...
Thanks in advance
Christian

**Reply by mithrendal** [reply]
This mail is well written. 👍🏻 Hope that they soon come back to you once their new Amiga takes over.

**Comment by chris70c** [comment]
Started working on joystick/keyboard settings UI

![controls_dark](https://user-images.githubusercontent.com/3858768/166077522-e8d232da-126e-4b65-89bc-910688239d61.png)

![controls_light](https://user-images.githubusercontent.com/3858768/166077525-9b0f4089-d928-4860-a354-30d022e5107b.png)

**Comment by chris70c** [comment]
Bringing back the good stuff :)

![vic_debug](https://user-images.githubusercontent.com/3858768/166126972-c464988d-d3b2-4595-8ae8-a59dac2ed6f9.png)

**Comment by dirkwhoffmann** [comment]
> Bringing back the good stuff :)

Very nice!!

I also started to backport stuff from vAmiga to VirtualC64. I.e., I am going to remove the "Disk" tab from the Inspector and move it to a separate window which can be opened via the drive menu (like "Df0 -> Inspect Disk..." in vAmiga). All other inspector tabs won't change much (if at all).

**Comment by chris70c** [comment]
https://www.neoartcr.com/debugger/

Try to drag one of the tabs (cia, sid, vic2) on the document... :)

**Comment by mithrendal** [comment]
>Try to drag one of the tabs (cia, sid, vic2) on the document... :)

hehe ... 😜 cool stuff ... I did as you said ... after that I did not figured it out to drag it back to where it initially was 🤔 ... maybe too hot and new stuff 🤤

**Reply by chris70c** [reply]
Just close the window, it will pop up back in the tabs...

**Comment by dirkwhoffmann** [comment]
> Try to drag one of the tabs (cia, sid, vic2) on the document... :)

Very cool. Works in Safari, too. 😎

**Comment by chris70c** [comment]
https://www.neoartcr.com/virtualc64_beta/

First rough implementation of  working debugger (if you split the tab the splitted one won't get updated because it seems that I can only set 1 inspection target at a time in the old vc64 version, I will soon update to the latest source code in dev).

If you're getting any error while loading a game, delete the old local storage indexed database please.

Updated to latest master branch code, now all the debugger tabs can work at the same time :)

**Comment by mithrendal** [comment]
Hi @chris70c hope you are doing well ! 

just profiled your 2.1b1 2022-05-20 version of vAmiga Online...

emulation is (except audio processor) executed in main ui thread ... it is not anymore executed in a worker thread ...  

<img width="505" alt="image" src="https://user-images.githubusercontent.com/17108995/206803749-76ff1b04-4f36-4c7f-a663-943f18fc5a79.png">

haven't you not executed the emulation in a worker thread? or did I remember that wrong ...

if you changed it back from worker to main thread then what was the reason / benefits for that move ? Is it maybe a compatibility reason ... (I have an experimental worker version online which does run quite well except iPhone/iPad 🙈 )

**Reply by chris70c** [reply]
Hi there, all well here, just didn't have much time to work on vAmiga lately...
Emulation was always on main thread, separate thread is for the audio backend, the main difference from your version is that I'm using shared memory between wasm and js so I don't have to copy the audio buffer nor the screen buffer I just get pointers from C and use them to send the appropriate buffer to both webgl and audio backend.
